include/boost/beast/websocket/impl/error.ipp

```diff
@@ -11,71 +11,112 @@
+    case error::no_connection:          return "The WebSocket handshake Connection field is missing";
+    case error::no_connection_upgrade:  return "The WebSocket handshake Connection field is missing the upgrade token";
+    case error::no_upgrade:             return "The WebSocket handshake Upgrade field is missing";
+    case error::no_upgrade_websocket:   return "The WebSocket handshake Upgrade field is missing the websocket token";
+    case error::no_sec_key:             return "The WebSocket handshake Sec-WebSocket-Key field is missing";
+    case error::bad_sec_key:            return "The WebSocket handshake Sec-WebSocket-Key field is invalid";
+    case error::no_sec_version:         return "The WebSocket handshake Sec-WebSocket-Version field is missing";
+    case error::bad_sec_version:        return "The WebSocket handshake Sec-WebSocket-Version field is invalid";
+    case error::no_sec_accept:          return "The WebSocket handshake Sec-WebSocket-Accept field is missing";
+    case error::bad_sec_accept:         return "The WebSocket handshake Sec-WebSocket-Accept field is invalid";
+    case error::upgrade_declined:       return "The WebSocket handshake was declined by the remote peer";
 
-    error_condition
-    default_error_condition(int ev) const noexcept override
-    {
-        return error_condition(ev, *this);
+    case error::bad_opcode:             return "The WebSocket frame contained an illegal opcode";
+    case error::bad_data_frame:         return "The WebSocket data frame was unexpected";
+    case error::bad_continuation:       return "The WebSocket continuation frame was unexpected";
+    case error::bad_reserved_bits:      return "The WebSocket frame contained illegal reserved bits";
+    case error::bad_control_fragment:   return "The WebSocket control frame was fragmented";
+    case error::bad_control_size:       return "The WebSocket control frame size was invalid";
+    case error::bad_unmasked_frame:     return "The WebSocket frame was unmasked";
+    case error::bad_masked_frame:       return "The WebSocket frame was masked";
+    case error::bad_size:               return "The WebSocket frame size was not canonical";
+    case error::bad_frame_payload:      return "The WebSocket frame payload was not valid utf8";
+    case error::bad_close_code:         return "The WebSocket close frame reason code was invalid";
+    case error::bad_close_size:         return "The WebSocket close frame payload size was invalid";
+    case error::bad_close_payload:      return "The WebSocket close frame payload was not valid utf8";
     }
+}
 
-    bool
-    equivalent(int ev,
-        error_condition const& condition
-            ) const noexcept override
+template<class>
+error_condition
+error_codes::
+default_error_condition(error e) const noexcept
+{
+    switch(e)
     {
-        return condition.value() == ev &&
-            &condition.category() == this;
-    }
+    default:
+    case error::closed:
```

Comment:
The fallthroughs here might generate a lot of warnings with `-Wimplicit-fallthrough` enabled.

Reply:
`-Wimplicit-fallthrough` already generates a lot of noise for the existing codebase. I [proposed](https://github.com/boostorg/beast/pull/956/commits/1a3c8d2856ddfdca14c544a7bf860788be0489e2) adding `-Wno-implicit-fallthrough` to the CMake file, and I assumed it was already set in the Jamfile, for the examples. I'd suggest addressing that separately and globally.

Reply:
That sounds to me like a defect in gcc or clang, these should not be producing warnings. Are you using a release version of the compiler? What version are you on where you see these warnings?

Reply:
gcc-7.2.0 mingw-w64 build in msys. According to [the implementer of the warning](https://developers.redhat.com/blog/2017/03/10/wimplicit-fallthrough-in-gcc-7/), it's been implemented only now that we have `[[fallthrough]]` to annotate deliberate switch fallthrough, and it's part of `-Wextra` in gcc 7. It might be considered a defect that it's active with `-std=c++11`, I guess, but since gcc has attributes to indicate deliberate fallthrough on all C++ levels, I can see the rationale here.

Assuming all target compilers correctly ignore unknown `[[]]` attributes (need to check this...), then the choice is either document that you need `-Wno-implicit-fallthrough` with gcc 7 or later, get gcc policy changed, or add `[[gnu::fallthrough]]` or some recognised comment to the falling-through lines.

Reply:
I've already dealt with the fallthrough warnings throughout Beast by using the `BOOST_FALLTHROUGH` macro. But I don't think the following code should cause a warning:
```
switch(n)
{
case 0:
case 1:
    std::cout << "low\n";
    break;
default:
    std::cout << "high\n";
break;
}
```

Or should it?

Reply:
Why wouldn't it? It's implicitly falling through from `case 0` to `case 1`. Do you intend that because `case  0` labels a null statement, the warning should treat it as labelling the same statement as `case 1`, and not complain? It's not a bad case to make, although the counter-position that comes to mind is that if that null statement was only because a macro had disappeared other statements, it's still carrying the same risk for the reader as the more-obvious implicit fallthrough.

Reply:
This particular case should not generate any warnings:
```cpp
switch (n) {
case 0:
case 1:
    /* code */
}
```
I wouldn't use BOOST_FALLTHROUGH for these cases.

Reply:
I haven't checked if this construct is actually the cause of the surprise warnings in gcc 7, but if it is, then maybe worth raising with gcc? A quick look at the discussion in https://gcc.gnu.org/bugzilla/show_bug.cgi?id=7652 suggests it's not supposed to warn in this case ("empty body"), which is [apparently consistent with clang's behaviour](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=7652#c13)